src/utils/utils.rs

```diff
@@ -593,6 +593,25 @@ pub fn toolchain_sort<T: AsRef<str>>(v: &mut Vec<T>) {
     });
 }
 
+fn copy_and_delete<'a, N>(
+    name: &'static str,
+    src: &'a Path,
+    dest: &'a Path,
+    notify_handler: &'a dyn Fn(N),
+) -> Result<()>
+where
+    N: From<Notification<'a>>,
+{
+    // https://github.com/rust-lang/rustup/issues/1239
+    // This uses std::fs::copy() instead of the faster std::fs::rename() to
+    // avoid cross-device link errors.
+    if src.is_file() {
+        copy_file(src, dest).and(remove_file(name, src))
+    } else {
+        copy_dir(src, dest, notify_handler).and(remove_dir(name, src, notify_handler))
```

Comment:
remove_dir sends a notification that a directory is being removed; thats sensible for e.g. removing an installed toolchain, but here I'm not sure we want to do that, since we're emulating rename; I think calling raw::remove_dir, or with the modification I suggest above, remove_dir_all::remove_dir_all(path) directly, would be better